Add Helm & Operator instructions and cleanup Cluster Checks instructions (#13441)
DataDog/documentation: content/en/agent/cluster_agent/clusterchecks.md

Metrics, events, and service checks collected by cluster checks are submitted without a hostname, as it is not relevant. A `cluster_name` tag is added, to allow you to scope and slice your data.

This strategy can be used if your infrastructure is configured for high availability (HA).

## Set up cluster check dispatching
The setup process involves enabling the dispatching ability in the Cluster Agent, as well as ensuring the Agents are prepared to receive configurations from the `clusterchecks` provider. Once this is done configurations can be passed to the Cluster Agent through mounted config files or through Kubernetes Service Annotations.

{{< tabs >}}
{{% tab "Helm" %}}

> This is enabled by default in the Helm deployment of the Cluster Agent, with respect to the `datadog.clusterChecks.enabled` configuration key. The configuration can be set as:

Reviewer: ```suggestion
Cluster check dispatching is enabled by default in the Helm deployment of the Cluster Agent through the `datadog.clusterChecks.enabled` configuration key:
```

```yaml
datadog:
  clusterChecks:
    enabled: true
  # (...)
clusterAgent:
  enabled: true
  # (...)
```

> This will enable the Cluster Check setup in the Cluster Agent and allow it to process configurations from the Kubernetes Service Annotations (`kube_services`).

Reviewer: ```suggestion
This enables the cluster check setup in the Cluster Agent and allows it to process configurations from the Kubernetes Service Annotations (`kube_services`).
```

{{% /tab %}}
{{% tab "Operator" %}}

> This can be enabled in the Operator deployment of the Cluster Agent with respect to the `clusterAgent.config.clusterChecksEnabled` configuration key. The configuration can be set as:

Reviewer: ```suggestion
Cluster check dispatching is enabled in the Operator deployment of the Cluster Agent by using the `clusterAgent.config.clusterChecksEnabled` configuration key:
```

```yaml
apiVersion: datadoghq.com/v1alpha1
kind: DatadogAgent
metadata:
  name: datadog
spec:
  # (...)
  clusterAgent:
    config:
      clusterChecksEnabled: true
```

> This will enable the Cluster Check setup in the Cluster Agent and allow it to process configurations from the Kubernetes Service Annotations (`kube_services`).

Reviewer: ```suggestion
This enables the cluster check setup in the Cluster Agent and allows it to process configurations from the Kubernetes Service Annotations (`kube_services`).
```

{{% /tab %}}
{{% tab "Daemonset" %}}
### Cluster Agent

This feature requires a running [Cluster Agent][1]. Once that is running make the following changes to the Cluster Agent Deployment:

1. Set the environment variable `DD_CLUSTER_CHECKS_ENABLED` to `true`.
2. Pass your cluster name as `DD_CLUSTER_NAME`. To help you scope your metrics, Datadog injects your cluster name as a `cluster_name` instance tag to all configurations.
3. If the service name is different from the default `datadog-cluster-agent`, ensure the `DD_CLUSTER_AGENT_KUBERNETES_SERVICE_NAME` environment variable reflects the service name.
4. To enable the Cluster Agent to process configurations from the Kubernetes Service Annotations set **both** `DD_EXTRA_CONFIG_PROVIDERS` and `DD_EXTRA_LISTENERS` environment variables to `kube_services`.

### Agent

Enable the `clusterchecks` configuration provider on the Datadog **Node** Agent. This can be done in two ways:

> - (Recommended) By setting the `DD_EXTRA_CONFIG_PROVIDERS` environment variable in your Agent DaemonSet. This takes a space separated string if you have multiple values:

Reviewer: ```suggestion
- **Recommended**: By setting the `DD_EXTRA_CONFIG_PROVIDERS` environment variable in your Agent DaemonSet. This takes a space-separated string if you have multiple values:
```

          polling: true
    ```

> [1]: /agent/cluster_agent/setup/

Reviewer: Did you mean to delete the "Restart the Agent" line? if so, you can also remove the link URL (line 90)

Author: Yes meant to remove that, that URL link is also referenced earlier too on line 65 so kept it for that.

{{% /tab %}}
{{< /tabs >}}

> **Note**: When running Cluster Checks the metrics reported by the Agent are not linked to a given hostname. As these are meant to be cluster centric metrics, and not necessarily host based metrics. As a result these metrics will also not inherit any host-level tags associated with that host. Such as those inherited from a cloud provider or added by the Agent's `DD_TAGS` environment variable. To add tags to cluster check metrics use the `DD_CLUSTER_CHECKS_EXTRA_TAGS` environment variable.

Reviewer: ```suggestion
**Note**: With cluster checks, the metrics reported by the Agent are not linked to a given hostname because they are meant to be cluster-centric metrics and not necessarily host-based metrics. As a result, these metrics do not inherit any host-level tags associated with that host, such as those inherited from a cloud provider or added by the Agent's `DD_TAGS` environment variable. To add tags to cluster check metrics, use the `DD_CLUSTER_CHECKS_EXTRA_TAGS` environment variable.
```

> The [Datadog Helm Chart][3] and the [Datadog Operator][4] additionally offer the possibility to deploy Cluster Check Runners. Which are a Deployment for a small set of Datadog Agents configured to run these dispatched cluster checks only, as opposed to dispatching these to the normal node-based Agents.

Reviewer: ```suggestion
The [Datadog Helm Chart][3] and the [Datadog Operator][4] additionally offer the possibility to deploy cluster check runners, which are a deployment for a small set of Datadog Agents configured to run these dispatched cluster checks only, as opposed to dispatching these to the normal node-based Agents.
```

### Advanced dispatching

The Cluster Agent can be configured to use an advanced dispatching logic for cluster checks, which takes into account the execution time and metric samples from check instances. This logic enables the Cluster Agent to optimize dispatching and distribution between cluster check runners.

#### Cluster Agent setup

In addition to the steps mentioned in the Cluster Agent Setup section, you must set the `DD_CLUSTER_CHECKS_ADVANCED_DISPATCHING_ENABLED` environment variable to `true`.

#### Agent setup

> The following environment variables are required to configure the node Agents (or Cluster Check Runners) to expose their check stats. The stats are consumed by the Cluster Agent and are used to optimize the cluster checks' dispatching logic.

Reviewer: ```suggestion
The following environment variables are required to configure the node Agents (or cluster check runners) to expose their check stats. The stats are consumed by the Cluster Agent and are used to optimize the cluster checks' dispatching logic.
```

        fieldRef:
          fieldPath: status.podIP
```
### Custom checks
Running [custom Agent checks][5] as cluster checks is supported, as long as all node-based Agents are able to run the check. This means your custom check code:

- Must be installed on all node-based Agents where the `clusterchecks` config provider is enabled.
- Must **not** depend on local resources that are not accessible to all Agents.

## Setting up check configurations

### Configuration from static configuration files

> When the URL or IP of a given resource is constant (eg. external service endpoint, public URL, etc.), a static configuration can be passed to the Cluster Agent as YAML files. The file name convention and syntax are the same as the static configurations on the node-based Agent. With the **required** addition of the `cluster_check: true` line.

Reviewer: ```suggestion
When the URL or IP of a given resource is constant (for example, an external service endpoint or a public URL), a static configuration can be passed to the Cluster Agent as YAML files. The file name convention and syntax are the same as the static configurations on the node-based Agent, with the **required** addition of the `cluster_check: true` line.
```

{{< tabs >}}
{{% tab "Helm" %}}

> With Helm these configuration files can be created within the `clusterAgent.confd` section. **Note**: This is separate from the `datadog.confd` section where the files are created in the node-based Agents. The `<INTEGRATION_NAME>` must exactly match the desired integration check you want to run.

Reviewer: ```suggestion
With Helm, these configuration files can be created within the `clusterAgent.confd` section. **Note**: This is separate from the `datadog.confd` section where the files are created in the node-based Agents. The `<INTEGRATION_NAME>` must exactly match the desired integration check you want to run.
```

```yaml
#(...)
clusterAgent:
  confd:
    <INTEGRATION_NAME>.yaml: |-
      cluster_check: true
      init_config:
        - <INIT_CONFIG>
      instances:
        - <INSTANCES_CONFIG>
```
{{% /tab %}}
{{% tab "Daemonset" %}}

> With the manual approach you must create a `ConfigMap` storing the desired static configuration files and mount that into the corresponding `/conf.d` file of the Cluster Agent container. This follows the same approach for [mounting ConfigMaps into the Agent container][1]. For example:

Reviewer: ```suggestion
With the manual approach you must create a `ConfigMap` to store the desired static configuration files, and mount it into the corresponding `/conf.d` file of the Cluster Agent container. This follows the same approach for [mounting ConfigMaps into the Agent container][1]. For example:
```

```yaml
kind: ConfigMap
apiVersion: v1
metadata:
  name: "<NAME>-config-map"
data:
  <INTEGRATION_NAME>-config: |-
    cluster_check: true
    init_config:
      <INIT_CONFIG>
    instances:
      <INSTANCES_CONFIG>
```

Then in the manifest for the Cluster Agent Deployment define the `volumeMounts` and `volumes` with respect to your `ConfigMap` and the corresponding key of your data.

```yaml
        volumeMounts:
          - name: <NAME>-config-map
            mountPath: /conf.d/
            # (...)
      volumes:
        - name: <NAME>-config-map
          configMap:
            name: <NAME>-config-map
            items:
              - key: <INTEGRATION_NAME>-config
                path: <INTEGRATION_NAME>.yaml
          #(...)
```

> This should result in a file in the `/conf.d/` directory of the Cluster Agent with respect to your desired integration. For example: `/conf.d/mysql.yaml` or `/conf.d/http_check.yaml`.

Reviewer: ```suggestion
This creates a file in the `/conf.d/` directory of the Cluster Agent corresponding to the integration. For example: `/conf.d/mysql.yaml` or `/conf.d/http_check.yaml`.
```

[1]: /agent/kubernetes/integrations/?tab=configmap#configuration
{{% /tab %}}
{{< /tabs >}}

#### Example: MySQL check on an externally hosted database

> After setting up an externally hosted database, such as CloudSQL or RDS, and a corresponding [Datadog user][6] to access the database, mount a `/conf.d/mysql.yaml` file in the Cluster Agent container with the following content:

Reviewer: ```suggestion
After you set up an externally hosted database, such as CloudSQL or RDS, and a corresponding [Datadog user][6] to access the database, mount a `/conf.d/mysql.yaml` file in the Cluster Agent container with the following content:
```

```yaml
cluster_check: true
init_config:
instances:
    - server: "<PRIVATE_IP_ADDRESS>"
      port: 3306
      user: datadog
      pass: "<YOUR_CHOSEN_PASSWORD>"
```

#### Example: HTTP_Check on an external URL

If there is a URL you would like the to perform an [HTTP check][7] against once per cluster, mount a `/conf.d/http_check.yaml` file in the Cluster Agent container with the following content:

```yaml
cluster_check: true
init_config:
instances:
    - name: "<EXAMPLE_NAME>"
      url: "<EXAMPLE_URL>"
```

> ### Configuration from kubernetes service annotations

Reviewer: ```suggestion
### Configuration from Kubernetes service annotations
```

        run: my-nginx
```

In addition, each pod should be monitored with the [NGINX check][9], as it enables the monitoring of each worker as well as the aggregated service.

## Troubleshooting

Due to the distributed nature of cluster checks, troubleshooting them is a bit more involved. The following sections explain the dispatching process and the associated troubleshooting commands.

### Kubernetes: find the leader Cluster Agent

> When leader election is enabled, only the leader serves Cluster Check configurations to the node-based Agents. If you only have 1 replica of the Cluster Agent pod running, that will be the leader. Otherwise, you can identify the name of the leader in the `datadog-leader-election` `ConfigMap`:

Reviewer: ```suggestion
When leader election is enabled, only the leader serves Cluster Check configurations to the node-based Agents. If only one replica of the Cluster Agent pod is running, it is the leader. Otherwise, you can identify the name of the leader in the `datadog-leader-election` `ConfigMap`:
```
